### PR TITLE
Detect sensitive terms in queries

### PR DESCRIPTION
**Pull request by olaughter**
This applies existing logic around sensitive search terms that currently is applied during opensearch queries and adds it for vespa queries.

### CODE REVIEW

**Review by harrisonpim — COMMENTED** [review]
The logic for detecting sensitive terms and making the switch looks great to me!

As far as I can tell, this PR will turn a query for a set of sensitive search terms into an *exact match* query. The current exact match implementation is really quite strict! Stemming's disabled on the search terms, their order needs to be respected, etc

Rather than reverting to a strict exact match, I think the previous implementation just turned off the KNN section of the query, while still allowing the standard lexical matches with tokens out of order, stemmed forms, etc. 

I think implementing that here would require a bunch of extra work in the DAL, so I just wanted to check what our current intent is before approving/pushing back!

**Review by kdutia — COMMENTED** [review]
agreed with @harrisonpim - i think to implement this we need to turn off KNN, rather than fall back to exact match. i'll catch up with him in a bit around how much work this is 

Also, good catch with the fix around string of sensitive terms!

**Comment by olaughter** [comment]
Thanks @kdutia & @harrisonpim, sounds like a separate ranking profile would need defining separately in the vespa application's schemas?

https://github.com/climatepolicyradar/navigator-infra/tree/1731056e29c887bdfe638d8824084b56f4208e26/vespa/navigator_app/schemas

**Comment by kdutia** [comment]
@olaughter i'm hoping we just need a different `where` subclause in the query [here](https://github.com/climatepolicyradar/data-access/blob/8ac8b0fc8c733d58be0b9d6c1f4a8cc56ab196cf/src/cpr_data_access/vespa.py#L113-L128), but will defer to @harrisonpim (or maybe neither of us know, and we need to try it out. i'm still not 100% clear on what's defined in the query  vs schema)

**Comment by harrisonpim** [comment]
Yeah, as far as I can tell this will require a new ranking profile, using `bm25()` as in the `hybrid` profile but without the `closeness()` term

**Comment by THOR300** [comment]
> Yeah, as far as I can tell this will require a new ranking profile, using `bm25()` as in the `hybrid` profile but without the `closeness()` term

@harrisonpim trying to understand why this would be the case as the actual yql query should be the same as far as I'm aware?

**Comment by olaughter** [comment]
@kdutia @harrisonpim like this?
https://github.com/climatepolicyradar/navigator-infra/pull/316

**Comment by kdutia** [comment]
@olaughter i gave that a tick

**Comment by olaughter** [comment]
Thanks Both, I've also updated the logic here and will test its working once that ranking profile goes live

**Comment by olaughter** [comment]
Update on that, I fixed a syntax issue, a rare bug and also [written a test for this query type](https://github.com/climatepolicyradar/navigator-infra/pull/319), Once this Pr gets an approve I'll deploy and deploy this in the backend

**Review by kdutia — APPROVED** [review]
lgtm 🎉

File: src/cpr_data_access/utils.py
Comment (THOR300) [inline src/cpr_data_access/utils.py:28]: image

Just to be clear!

File: src/cpr_data_access/utils.py
Comment (THOR300) [inline src/cpr_data_access/utils.py:22]: `set(map(str.lower, sensitive_terms)).intersection(set(text.lower().split()))`?

File: src/cpr_data_access/search_adaptors.py
Comment (THOR300) [inline src/cpr_data_access/search_adaptors.py:96]: What does hybrid no closeness mean for a ranking profile, can't see anything in the docs?

**Review by THOR300 — APPROVED** [review]
LGTM

File: src/cpr_data_access/utils.py
Comment (olaughter) [reply src/cpr_data_access/utils.py:28]: Good spot, although I think this should be okay as these words are coming out of the tsv file. I'll double check

Comment (olaughter) [reply src/cpr_data_access/utils.py:28]: Interestingly it turns out some of those words have trailing spaces, although the load function cleans them up, so seems we only end up with phrases that don't have extra spaces

File: src/cpr_data_access/utils.py
Comment (olaughter) [reply src/cpr_data_access/utils.py:22]: I'm not sure we can use `text.lower().split()` here because some of the sensitive terms are multiple words

Comment (olaughter) [reply src/cpr_data_access/utils.py:22]: Just profiled both options and seems set intersection is a lot slower as well, unfortunately

File: src/cpr_data_access/search_adaptors.py
Comment (olaughter) [reply src/cpr_data_access/search_adaptors.py:96]: Heres where the definition is added: https://github.com/climatepolicyradar/navigator-infra/pull/316

**Pull request merged**
